File: day02.kt
package day02

import java.io.File

enum class Kind {
    FORWARD, UP, DOWN
}

data class Command(
    val kind: Kind,
    val arg: Int
)

fun String.toCommand(): Command {
    val (kindStr, argStr) = this.split(' ')
    val kind: Kind = when (kindStr) {
        "forward" -> Kind.FORWARD
        "up" -> Kind.UP
        "down" -> Kind.DOWN
        else -> throw IllegalArgumentException("unknown command type: $kindStr")
    }
    val arg: Int = argStr.toInt()
    return Command(kind, arg)
}

data class State(
    var horizontal: Int = 0,
    var depth: Int = 0,
    var aim: Int = -1
) {
    fun applyCommand(cmd: Command) {
        if (this.aim == -1) {
            when (cmd.kind) {
                Kind.FORWARD -> this.horizontal += cmd.arg
                Kind.DOWN -> this.depth += cmd.arg
                Kind.UP -> this.depth -= cmd.arg
            }
        } else {
            when (cmd.kind) {
                Kind.FORWARD -> {
                    this.horizontal += cmd.arg
                    this.depth += this.aim * cmd.arg
                }
                Kind.DOWN -> this.aim += cmd.arg
                Kind.UP -> this.aim -= cmd.arg
            }
        }
    }

}

fun part1(commands: List<Command>): Int {
    val state = State(0, 0)
    commands.forEach { state.applyCommand(it) }
    return state.horizontal * state.depth
}

fun part2(commands: List<Command>): Int {
    val state = State(0, 0, 0)
    commands.forEach { state.applyCommand(it) }
    return state.horizontal * state.depth
}

fun main() {
    val data = File("../inputs/day02.txt").readLines().map { it.toCommand() }
    val answer1 = part1(data)
    println("Answer 1: $answer1") // 2019945
    val answer2 = part2(data)
    println("Answer 2: $answer2") // 1599311480
}

File: day02.py
from dataclasses import dataclass
from pathlib import Path
import attr


@attr.s
class Command:
    kind: attr.ib(type=str)
    arg: attr.ib(type=int, converter=int)

    @classmethod
    def from_string(cls, line: str):
        (kind, arg) = line.split(' ')
        return Command(kind, arg)


@dataclass
class State:
    horizontal: int
    depth: int
    aim: int = None

    def apply_command(self, cmd: Command) -> 'State':
        if self.aim is None:
            if cmd.kind == "forward":
                self.horizontal = self.horizontal + cmd.arg
            if cmd.kind == "down":
                self.depth = self.depth + cmd.arg
            if cmd.kind == "up":
                self.depth = self.depth - cmd.arg
        else:
            if cmd.kind == "forward":
                self.horizontal = self.horizontal + cmd.arg
                self.depth = self.depth + self.aim * cmd.arg
            if cmd.kind == "down":
                self.aim = self.aim + cmd.arg
            if cmd.kind == "up":
                self.aim = self.aim - cmd.arg
        return self


def part1(cmds: list[Command]) -> int:
    state = State(0, 0)
    for cmd in cmds:
        state.apply_command(cmd)
    return state.depth * state.horizontal


def part2(cmds: list[Command]) -> int:
    state = State(0, 0, 0)
    for cmd in cmds:
        state.apply_command(cmd)
    return state.depth * state.horizontal


if __name__ == '__main__':
    data = Path('inputs/day02.txt').read_text(encoding='utf-8')
    split_data = data.splitlines()
    commands = [Command.from_string(line) for line in split_data]
    print('Answer 1:', part1(commands))  # 2019945
    print('Answer 2:', part2(commands))  # 1599311480
